doc/devel/contributing.rst

```diff
@@ -349,6 +349,23 @@ C/C++ extensions
   docstrings, and the Numpydoc format is well understood in the
   scientific Python community.
 
+New public APIs
+---------------
+
+Public APIs added to Matplotlib are bound by backwards compatibility, and
+therefore cumbersome to change.  Therefore, when adding new functionality, try
+to add as little public API as possible.  In particular, make sure that helper
+methods and internal attributes are private (i.e., start with an underscore).
+Remember that any non-private method can be directly called by the end user,
+and any non-private attribute can be directly set!
```

Comment:
I'd probably leave out this sentence. It is extraneous and doesn't really explain why this is a "bad thing" (it can also confuse some contributors because they'll notice that they can do exactly the same thing with private methods and attributes).

Reply:
I'm not particularly in love with the wording here, but I do want to stress that point extremely strongly: if an attribute is public, the implementation should be robust against the end user modifying it; otherwise, make it private, or hide it behind a getter or a property.
If you have a better wording I'm all ears :)

Reply:
Or add a  "And therefore once we have released that API the behavior, arguments, and names can not be changed without going through a multi-version deprecation process." ?